app/src/main/java/uk/gov/onelogin/network/auth/AuthCodeExchange.kt

```diff
@@ -0,0 +1,91 @@
+package uk.gov.onelogin.network.auth
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.Companion.PRIVATE
+import com.google.gson.Gson
+import dagger.hilt.android.qualifiers.ApplicationContext
+import io.ktor.client.request.post
+import io.ktor.client.statement.bodyAsText
+import io.ktor.http.ContentType
+import io.ktor.http.HttpStatusCode
+import io.ktor.http.Url
+import io.ktor.http.contentType
+import uk.gov.onelogin.R
+import uk.gov.onelogin.network.auth.response.TokenResponse
+import uk.gov.onelogin.network.http.IHttpClient
+import uk.gov.onelogin.network.utils.IOnlineChecker
+import javax.inject.Inject
+
+class AuthCodeExchange @Inject constructor(
+    @ApplicationContext
+    private val context: Context,
+    private val httpClient: IHttpClient,
+    private val onlineChecker: IOnlineChecker
+): IAuthCodeExchange {
+    @VisibleForTesting(otherwise = PRIVATE)
+    private var redirectUrl: Url = Url(
+        context.resources.getString(
+            R.string.webBaseUrl,
+            context.resources.getString(R.string.webRedirectEndpoint)
+        )
+    )
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    private var url: Url = Url(
+        context.resources.getString(
+            R.string.apiBaseUrl,
+            context.resources.getString(R.string.tokenExchangeEndpoint)
+        )
+    )
+
+    override suspend fun exchangeCode(code: String): TokenResponse {
+        if (code.isEmpty()) {
+            throw AuthCodeExchangeCodeArgError("Code should be a non-empty string")
+        }
+
+        if (!onlineChecker.isOnline()) {
+            throw AuthCodeExchangeOfflineError("The device appears to be offline")
+        }
+
+        val response = httpClient.client().post(url) {
+            url {
+                parameters.apply {
+                    append("grant_type", "authorization_code" )
+                    append("code", code)
+                    append("redirect_uri", redirectUrl.toString())
+                }
+            }
+            contentType(
+                ContentType.Application.FormUrlEncoded
+            )
+        }
+
+        return when {
```

Comment:
An open `when` block like this is essentially a fancy nested `if` statement or a JS-ish `switch`, but cleaner. An open when takes a boolean expression and executes the first truthy condition. The equivalent in JS would look something like this:

```javascript
switch (true) {
  case response.status >= 500: {
    throw new AuthCodeExchangeServerError(
        `Server Error received - ${response.status} - ${response.body}`
    )
  }
  ...
  default: {
    return JSON.parse(response.body)
  }
}
```

This when is also returning, which is a nice feature of Kotlin whereby you can return an `if`/`when` etc so you don't miss a `return` statement 😉